File: rbq_sdk/python/example/sport/rbq_high_level.py
"""Python port of rbq_sdk/cpp/example/src/rbq_high_level.cpp.

Publishes HighLevelCommand at 50 Hz and switchControlMode=true at startup,
drives gait via terminal input, and forwards joystick sticks to velocity
setpoints. Matches the C++ example's behavior 1:1.

Usage:
    python3 rbq_high_level.py <networkInterface> [joystickPath]
"""

import array
import errno
import fcntl
import glob
import os
import signal
import struct
import sys
import threading
import time

from rbq_sdk_py.core.channel import ChannelFactoryInitialize, ChannelPublisher

# Locate the sibling `example/idl/` package so `_rbq_idl` resolves when the
# example is launched directly (python3 rbq_high_level.py ...).
sys.path.insert(
    0,
    os.path.normpath(
        os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "idl")
    ),
)
from _rbq_idl import HighLevelCommand_, Bool_  # noqa: E402


APP_NAME = "rbq_high_level"


# ---------------------------------------------------------------------------
# Gait table (mirrors GAIT_STATE in rbq_high_level.cpp)
# ---------------------------------------------------------------------------

GAIT_OPTIONS = [
    ("sitting",         0),
    ("standing",        1),
    ("aiming",          2),
    ("trotting",        3),
    ("trot_stairs",     4),
    ("waving",          5),
    ("trot_running",    6),
    ("docking",         10),
    ("rl_trot",         30),
    ("rl_front_walk",   31),
    ("rl_left_walk",    33),
    ("rl_right_walk",   34),
    ("rl_bound",        35),
    ("rl_pace",         36),
    ("rl_pronk",        37),
    ("rl_3leg_hr",      38),
    ("rl_3leg_hl",      39),
    ("rl_3leg_fl",      41),
    ("rl_trot_vision", 42),
    ("rl_trot_run",     45),
    ("rl_silent",       46),
    ("rl_stairs",       47),
    ("rl_end",          80),
]

DEADZONE = 0.12
MAX_VX = 1.2        # m/s
MAX_VY = 0.5        # m/s
# QuadWalk treats HighLevelCommand.omega_z as deg/s and multiplies by D2R
# internally; 60 deg/s ≈ 1.05 rad/s yaw rate.
MAX_VYAW = 60.0     # deg/s

DEFAULT_GAIT_STATE = 1  # STANDING


# ---------------------------------------------------------------------------
# Joystick (linux /dev/input/js* raw API)
# ---------------------------------------------------------------------------

JS_EVENT_BUTTON = 0x01
JS_EVENT_AXIS = 0x02
JS_EVENT_INIT = 0x80

JSIOCGNAME_128 = (2 << 30) | (128 << 16) | (ord("j") << 8) | 0x13

AXIS_COUNT = 8
BTN_COUNT = 16
AXIS_LEFT_X, AXIS_LEFT_Y, AXIS_RIGHT_X = 0, 1, 3


def deadzone(v: float, dz: float) -> float:
    if abs(v) < dz:
        return 0.0
    sign = 1.0 if v > 0 else -1.0
    return sign * (abs(v) - dz) / (1.0 - dz)


def list_js_nodes():
    return sorted(p for p in glob.glob("/dev/input/js*") if os.path.exists(p))


def joy_name(fd: int) -> str:
    buf = array.array("B", [0] * 128)
    try:
        fcntl.ioctl(fd, JSIOCGNAME_128, buf)
    except OSError:
        return ""
    return buf.tobytes().split(b"\0", 1)[0].decode("utf-8", errors="replace")


def name_looks_like_pad(name: str) -> bool:
    n = name.lower()
    return any(k in n for k in ("logitech", "gamepad", "x-box", "xbox", "wireless"))


def try_open(path: str) -> bool:
    try:
        fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
        os.close(fd)
        return True
    except OSError:
        return False


def pick_joy_path(argv):
    if len(argv) >= 3:
        return argv[2] if try_open(argv[2]) else None
    env = os.environ.get("JOY", "")
    if env and try_open(env):
        return env
    fallback = ""
    for p in list_js_nodes():
        try:
            fd = os.open(p, os.O_RDONLY | os.O_NONBLOCK)
        except OSError:
            continue
        try:
            nm = joy_name(fd)
        finally:
            os.close(fd)
        if name_looks_like_pad(nm):
            return p
        if not fallback:
            fallback = p
    if fallback:
        return fallback
    return "/dev/input/js0" if try_open("/dev/input/js0") else None


class Joystick:
    def __init__(self, path: str):
        self.path = path
        self.fd = -1
        self.ax = [0.0] * AXIS_COUNT
        self.btn = [False] * BTN_COUNT
        self.prev = [False] * BTN_COUNT

    def open(self) -> bool:
        try:
            self.fd = os.open(self.path, os.O_RDONLY | os.O_NONBLOCK)
        except OSError as ex:
            print(f"조이스틱 열기 실패: {self.path} ({ex})", file=sys.stderr)
            return False
        nm = joy_name(self.fd)
        if nm:
            print(f"조이스틱: {nm} ({self.path})")
        return True

    def close(self):
        if self.fd >= 0:
            os.close(self.fd)
            self.fd = -1

    def poll(self):
        while True:
            try:
                data = os.read(self.fd, 8)
            except OSError as ex:
                if ex.errno in (errno.EAGAIN, errno.EWOULDBLOCK):
                    break
                print("조이 read 오류", file=sys.stderr)
                break
            if len(data) < 8:
                break
            _t, value, typ, number = struct.unpack("<IhBB", data)
            k = typ & ~JS_EVENT_INIT
            if k == JS_EVENT_AXIS and number < AXIS_COUNT:
                self.ax[number] = max(-1.0, min(1.0, value / 32767.0))
            elif k == JS_EVENT_BUTTON and number < BTN_COUNT:
                self.btn[number] = value != 0

    def end_frame(self):
        for i in range(BTN_COUNT):
            self.prev[i] = self.btn[i]


# ---------------------------------------------------------------------------
# Shared state (replaces the C++ std::atomic globals)
# ---------------------------------------------------------------------------


class SharedState:
    def __init__(self):
        self.is_working = True
        self.target_gait = DEFAULT_GAIT_STATE
        self.gait_change_pending = False
        self.vel_x = 0.0
        self.vel_y = 0.0
        self.omega_z = 0.0
        # gait_change_pending must be consumed exactly once by the publisher
        # tick (mirrors std::atomic<bool>::exchange in the C++ source).
        self._gait_lock = threading.Lock()

    def consume_gait_pending(self) -> bool:
        with self._gait_lock:
            v = self.gait_change_pending
            self.gait_change_pending = False
            return v


# ---------------------------------------------------------------------------
# Threads
# ---------------------------------------------------------------------------


def control_loop(state: SharedState):
    """Publish HighLevelCommand_ at 50 Hz and SwitchControlMode=true at startup."""
    pub_cmd = ChannelPublisher("rt/rbq/cmd/high_level", HighLevelCommand_)
    pub_cmd.Init()

    # Without EXT_JOY=true the on-robot QuadWalk ignores HighLevelCommand
    # (see src/QuadWalk/main.cpp). Flip it on at startup.
    pub_mode = ChannelPublisher("rt/rbq/cmd/switch_control_mode", Bool_)
    pub_mode.Init()
    time.sleep(0.2)
    on = Bool_(data=True)
    pub_mode.Write(on)
    print("EXT_JOY mode: ON (switchControlMode = true)")

    msg = HighLevelCommand_()
    dt = 0.02  # 50 Hz
    next_tick = time.monotonic()
    while state.is_working:
        next_tick += dt

        msg.gait_state = int(state.target_gait)
        msg.gait_transition = state.consume_gait_pending()
        msg.vel_x = float(state.vel_x)
        msg.vel_y = float(state.vel_y)
        msg.omega_z = float(state.omega_z)
        pub_cmd.Write(msg)

        delay = next_tick - time.monotonic()
        if delay > 0:
            time.sleep(delay)
        else:
            next_tick = time.monotonic()

    # Final stop on exit.
    msg.gait_transition = False
    msg.vel_x = 0.0
    msg.vel_y = 0.0
    msg.omega_z = 0.0
    pub_cmd.Write(msg)


def joystick_loop(state: SharedState, joy: Joystick):
    """Poll joystick at 100 Hz and update velocity setpoints."""
    while state.is_working:
        joy.poll()
        lx = deadzone(joy.ax[AXIS_LEFT_X], DEADZONE)
        ly = deadzone(joy.ax[AXIS_LEFT_Y], DEADZONE)
        rx = deadzone(joy.ax[AXIS_RIGHT_X], DEADZONE)
        state.vel_x = -ly * MAX_VX
        state.vel_y = -lx * MAX_VY
        state.omega_z = -rx * MAX_VYAW
        joy.end_frame()
        time.sleep(0.01)


# ---------------------------------------------------------------------------
# Terminal UI
# ---------------------------------------------------------------------------


def print_help():
    print(f"\n {APP_NAME} Help")
    print(f"\n Usage: {APP_NAME} <networkInterface> [joystickPath]")
    print("\n Gait controls (terminal input):")
    print("   list           - list all gait options")
    print("   <name>         - select gait by name (e.g. standing)")
    print("   <id>           - select gait by id   (e.g. 1)")
    print("\n Motion controls (joystick):")
    print("   Left stick     - vel_x / vel_y")
    print("   Right stick X  - omega_z\n")


def convert_to_int(s: str):
    try:
        return int(s)
    except ValueError:
        return None


def terminal_handle(state: SharedState, current):
    try:
        line = input()
    except EOFError:
        state.is_working = False
        return
    line = line.strip()

    if line == "list":
        for name, gid in GAIT_OPTIONS:
            print(f"{name}, id: {gid}")
        return

    candidate_int = convert_to_int(line)
    for name, gid in GAIT_OPTIONS:
        if line == name or candidate_int == gid:
            current["name"] = name
            current["id"] = gid
            state.target_gait = gid
            state.gait_change_pending = True
            print(f"Gait: {name}, gait_state: {gid}")
            return


# ---------------------------------------------------------------------------
# Entrypoint
# ---------------------------------------------------------------------------


def main(argv):
    if len(argv) < 2:
        print(f"Usage: {argv[0]} <networkInterface> [joystickPath]")
        return -1

    state = SharedState()

    def signal_handler(sig, frame):
        print(f"Signal received: {sig}")
        state.is_working = False
        os._exit(sig)

    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGHUP, signal_handler)

    jpath = pick_joy_path(argv)
    if not jpath:
        print(
            "Joystick not detected on /dev/input/js*. "
            "Check if USB dongle is connected and joystick is on.",
            file=sys.stderr,
        )
        for p in list_js_nodes():
            print(f"  {p}", file=sys.stderr)
        return -1
    if len(argv) < 3:
        print(f"Joystick path: {jpath}")

    joy = Joystick(jpath)
    if not joy.open():
        return -1

    ChannelFactoryInitialize(0, argv[1])

    current = {"name": "standing", "id": DEFAULT_GAIT_STATE}

    print_help()
    print('Input "list" to list all gait options ...')

    t_joy = threading.Thread(
        target=joystick_loop, args=(state, joy), name="JoyLoop", daemon=True
    )
    t_ctrl = threading.Thread(
        target=control_loop, args=(state,), name="ControlLoop", daemon=True
    )
    t_joy.start()
    t_ctrl.start()

    try:
        while state.is_working:
            terminal_handle(state, current)
    except KeyboardInterrupt:
        state.is_working = False

    t_joy.join()
    t_ctrl.join()
    joy.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: rbq_sdk/cpp/example/src/rbq_high_level.cpp
#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <vector>
#include <limits>
#include <atomic>
#include <stdexcept>
#include <csignal>

#include <rbq_sdk/dds/ChannelFactory.hpp>
#include <rbq_sdk/dds/Publisher.hpp>
#include <rbq_sdk/idl/rbq/HighLevelCommand_.hpp>
#include <rbq_sdk/idl/ros2/Bool_.hpp>
#include <rbq_sdk/Joystick.hpp>
#include <rbq_sdk/Thread.hpp>

using namespace std;

enum GAIT_STATE : int8_t {
    SITTING              = 0,
    STANDING             = 1,
    AIMING               = 2,
    TROTTING             = 3,
    TROT_STAIRS          = 4,
    WAVING               = 5,
    TROT_RUNNING         = 6,
    DOCKING              = 10,
    RL_TROT              = 30,
    RL_FRONT_WALK        = 31,
    RL_LEFT_WALK         = 33,
    RL_RIGHT_WALK        = 34,
    RL_BOUND             = 35,
    RL_PACE              = 36,
    RL_PRONK             = 37,
    RL_3LEG_HR           = 38,
    RL_3LEG_HL           = 39,
    RL_3LEG_FL           = 41,
    RL_TROT_VISION       = 42,
    RL_TROT_RUN          = 45,
    RL_SILENT            = 46,
    RL_STAIRS            = 47,
    RL_END               = 80,
};

struct TestOption
{
    std::string name;
    int id;
};

const vector<TestOption> option_list =
    {
        {"sitting",              SITTING             },
        {"standing",             STANDING            },
        {"aiming",               AIMING              },
        {"trotting",             TROTTING            },
        {"trot_stairs",          TROT_STAIRS         },
        {"waving",               WAVING              },
        {"trot_running",         TROT_RUNNING        },
        {"docking",              DOCKING             },
        {"rl_trot",              RL_TROT             },
        {"rl_front_walk",        RL_FRONT_WALK       },
        {"rl_left_walk",         RL_LEFT_WALK        },
        {"rl_right_walk",        RL_RIGHT_WALK       },
        {"rl_bound",             RL_BOUND            },
        {"rl_pace",              RL_PACE             },
        {"rl_pronk",             RL_PRONK            },
        {"rl_3leg_hr",           RL_3LEG_HR          },
        {"rl_3leg_hl",           RL_3LEG_HL          },
        {"rl_3leg_fl",           RL_3LEG_FL          },
        {"rl_trot_vision",       RL_TROT_VISION      },
        {"rl_trot_run",          RL_TROT_RUN         },
        {"rl_silent",            RL_SILENT           },
        {"rl_stairs",            RL_STAIRS           },
        {"rl_end",               RL_END              },
};

// -1 is a real gait_state (CONTROL_OFF), so use INT_MIN as the parse-failure
// sentinel instead of -1 like sport_client does.
constexpr int kInvalidId = std::numeric_limits<int>::min();

constexpr float kDeadzone = 0.12f;
constexpr float kMaxVx    = 1.2f;   // m/s
constexpr float kMaxVy    = 0.5f;   // m/s
// in deg/s — 60 deg/s ~ 1.05 rad/s of yaw.
constexpr float kMaxVyaw  = 60.0f;   // deg/s

std::atomic<bool>   g_isWorking{true};
std::atomic<int8_t> g_targetGait{static_cast<int8_t>(STANDING)};
std::atomic<float>  g_velX{0.f};
std::atomic<float>  g_velY{0.f};
std::atomic<float>  g_omegaZ{0.f};

void signalHandler(int sig)
{
    std::cout << "Signal received: " << sig << "\n";
    g_isWorking.store(false);
    std::_Exit(sig);
}

int ConvertToInt(const std::string &str)
{
    try {
        return std::stoi(str);
    } catch (const std::invalid_argument &) {
        return kInvalidId;
    } catch (const std::out_of_range &) {
        return kInvalidId;
    }
}

void printHelp()
{
    std::cout << "\n " << APP_NAME << " Help\n";
    std::cout << "\n Usage: " << APP_NAME << " <networkInterface> [joystickPath]\n";
    std::cout << "\n Gait controls (terminal input):\n";
    std::cout << "   list           - list all gait options\n";
    std::cout << "   <name>         - select gait by name (e.g. standing)\n";
    std::cout << "   <id>           - select gait by id   (e.g. 1)\n";
    std::cout << "\n Motion controls (joystick):\n";
    std::cout << "   Left stick     - vel_x / vel_y\n";
    std::cout << "   Right stick X  - omega_z\n\n";
}

// Publishes the current gait + joystick-driven velocities at 50 Hz.
void* controlLoop(void*)
{
    timespec timeEnd;
    timespec timeNext;
    clock_gettime(CLOCK_REALTIME, &timeNext);

    rbq_msgs::msg::dds_::HighLevelCommand_ msg;
    rbq_sdk::Publisher<rbq_msgs::msg::dds_::HighLevelCommand_> pubHighLevelCmd("rt/rbq/cmd/high_level");

    rbq_sdk::Publisher<std_msgs::msg::dds_::Bool_> pubSwitchControlMode("rt/rbq/cmd/switch_control_mode");
    {
        std_msgs::msg::dds_::Bool_ on;
        on.data() = true;
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        pubSwitchControlMode.write(on);
    }

    while (g_isWorking.load()) {
        msg.gait_state()      = g_targetGait.load(std::memory_order_relaxed);
        msg.gait_transition() = true;
        msg.vel_x()           = g_velX.load(std::memory_order_relaxed);
        msg.vel_y()           = g_velY.load(std::memory_order_relaxed);
        msg.omega_z()         = g_omegaZ.load(std::memory_order_relaxed);
        pubHighLevelCmd.write(msg);

        clock_gettime(CLOCK_REALTIME, &timeEnd);
        rbq_sdk::Thread::timespec_add_us(&timeNext, 20 * 1000); // 50 Hz
        if (rbq_sdk::Thread::timespec_cmp(&timeEnd, &timeNext) > 0) {
            clock_gettime(CLOCK_REALTIME, &timeNext);
        } else {
            clock_nanosleep(CLOCK_REALTIME, TIMER_ABSTIME, &timeNext, NULL);
        }
    }

    return nullptr;
}

// Polls the joystick at 100 Hz and updates atomic velocity setpoints.
void* joystickLoop(void* arg)
{
    Joystick* joy = static_cast<Joystick*>(arg);
    while (g_isWorking.load()) {
        joy->Poll();
        const float lx = Deadzone(joy->Ax(AXIS_LEFT_X),  kDeadzone);
        const float ly = Deadzone(joy->Ax(AXIS_LEFT_Y),  kDeadzone);
        const float rx = Deadzone(joy->Ax(AXIS_RIGHT_X), kDeadzone);
        g_velX.store(-ly * kMaxVx,   std::memory_order_relaxed);
        g_velY.store(-lx * kMaxVy,   std::memory_order_relaxed);
        g_omegaZ.store(-rx * kMaxVyaw, std::memory_order_relaxed);
        joy->EndFrame();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    return nullptr;
}

class UserInterface
{
public:
    void terminalHandle()
    {
        std::string input;
        if (!std::getline(std::cin, input)) {
            g_isWorking.store(false);
            return;
        }

        if (input.compare("list") == 0) {
            for (const TestOption &option : option_list) {
                std::cout << option.name << ", id: " << option.id << "\n";
            }
            return;
        }

        for (const TestOption &option : option_list) {
            if (input.compare(option.name) == 0 || ConvertToInt(input) == option.id) {
                test_option_->id   = option.id;
                test_option_->name = option.name;
                g_targetGait.store(static_cast<int8_t>(option.id), std::memory_order_relaxed);
                std::cout << "Gait: " << test_option_->name << ", gait_state: " << test_option_->id << "\n";
                return;
            }
        }
    }

    TestOption *test_option_;
};

int main(int argc, char **argv)
{
    if (argc < 2) {
        std::cout << "Usage: " << argv[0] << " <networkInterface> [joystickPath]\n";
        return -1;
    }

    signal(SIGTERM, signalHandler);
    signal(SIGINT,  signalHandler);
    signal(SIGHUP,  signalHandler);

    const std::string jpath = PickJoyPath(argc, argv);
    if (jpath.empty()) {
        PrintNoJoystickHint();
        return -1;
    }
    if (argc < 3) {
        std::cout << "Joystick path: " << jpath << '\n';
    }

    Joystick joy(jpath);
    if (!joy.Open()) {
        return -1;
    }

    rbq_sdk::ChannelFactory::Instance().Init(0, argv[1]);

    TestOption test_option;
    test_option.id   = STANDING;
    test_option.name = "standing";
    g_targetGait.store(static_cast<int8_t>(STANDING));

    UserInterface user_interface;
    user_interface.test_option_ = &test_option;

    printHelp();
    std::cout << "Input \"list\" to list all gait options ...\n";

    std::thread joystickThread(joystickLoop, &joy);
    std::thread controlThread(controlLoop, nullptr);

    while (g_isWorking.load()) {
        user_interface.terminalHandle();
    }

    if (joystickThread.joinable()) joystickThread.join();
    if (controlThread.joinable())  controlThread.join();
    return 0;
}
